### ISSUE TITLE
Some methods missing

### ISSUE THREAD

**Issue by Ehesp** [jodinathan/js_bindings#24]
It seems as though some classes, for example `Headers` is missing some methods:

![image](https://user-images.githubusercontent.com/842078/214844882-bbd8e9ae-a098-4b64-9688-9592144e58eb.png)

```dart
class Headers {
  external factory Headers([dynamic init]);
}

extension PropsHeaders on Headers {
  void append(String name, String value) =>
      js_util.callMethod(this, 'append', [name, value]);

  void delete(String name) => js_util.callMethod(this, 'delete', [name]);

  @JS('get')
  @staticInterop
  String? mGet(String name) => js_util.callMethod(this, 'get', [name]);

  bool has(String name) => js_util.callMethod(this, 'has', [name]);

  @JS('set')
  @staticInterop
  void mSet(String name, String value) =>
      js_util.callMethod(this, 'set', [name, value]);
}
```

**Comment by Ehesp** [comment]
Hmm looks like it might be `iterable` methods not being supported/handled. If we can handle them in the build code (not sure where it is), I think the API could probably look like this:

```dart
Iterator keys() => js_util.callMethod(this, 'keys', []);
```

```dart
@js.JS()
@js.staticInterop
class Iterator {
  external factory Iterator();
}

extension PropsIterator on Iterator {
  IteratorNext next() => js_util.callMethod(this, 'next', []);
}

@js.anonymous
@js.JS()
class IteratorNext {
  external factory IteratorNext();
}

extension PropsIteratorNext on IteratorNext {
  bool get done => js_util.getProperty(this, 'done');
  dynamic get value => js_util.getProperty(this, 'value');
}
```

What do you think?

**Comment by Ehesp** [comment]
Hmm, you already have some iterator logic which extends JsArray. Not sure :D

**Comment by jodinathan** [comment]
There are different specifications even thought it is all JS.
JsArray, iterator etc are from Emacscript IIRC

I've created [an issue](https://github.com/jodinathan/js_bindings/issues/7) regarding this

Maybe we should hardcode the missing methods?

**Comment by Ehesp** [comment]
I think hardcoding would be an OK approach, maybe just some custom extensions which are exported. Given the amount of work involved and how little the apis change I think it's probably the route to fastest solution?

If you could do one in a branch to demonstrate how to best do it, I'm happy to add them in places too.

**Comment by jodinathan** [comment]
In the case of `Headers` we would need to check the IDL JSON files where it is defined if it extend something like `JsArray` or `JsMap` then check in the builder why it is not recognizing and implementing.  
For the sake of teaching the quirks I already found it in the `fetch.json` line 88. It has no inheritance, however, there is a `iterable` member that I was not sure if the builder takes that into any consideration. 
Digging into the code I've found that if there is a member called `iterable`, the object [should extend](https://github.com/jodinathan/js_bindings/blob/main/tool/package_js/build.dart#L216) `JsArray`.
You can check what is going on there.

About the remaining methods and properties of `JsArray`, `JsMap` or any other JS object, just take `JsArray` as example and add the stuff into the extension.

**Comment by Ehesp** [comment]
Thanks - I'm not entirely sure how this works, but it looks as though it could be the builder which isn't adding the methods to the IDL files?

I see what you mean here, however there's no name and it's also missing the various methods.

```json
{
            "type": "iterable",
            "idlType": [
              {
                "type": null,
                "extAttrs": [],
                "generic": "",
                "nullable": false,
                "union": false,
                "idlType": "ByteString"
              },
              {
                "type": null,
                "extAttrs": [],
                "generic": "",
                "nullable": false,
                "union": false,
                "idlType": "ByteString"
              }
            ],
            "arguments": [],
            "extAttrs": [],
            "readonly": false,
            "async": false
          }
```

**Comment by Ehesp** [comment]
FYI @jodinathan no clue if you're interested, but I'd be willing to support development of this package. My company is using it for a project - feel free to reach out on Twitter/email if interested!

**Comment by jodinathan** [comment]
> FYI @jodinathan no clue if you're interested, but I'd be willing to support development of this package. My company is using it for a project - feel free to reach out on Twitter/email if interested!

Hi @Ehesp I've emailed you

**Comment by jodinathan** [comment]
@Ehesp I am closing this issue as it should be fixed in 0.1.0.

If you find more stuff missing please create a new issue
